kindelia/src/main.rs

```diff
@@ -860,9 +861,33 @@ pub fn start_node<C: ProtoComm + 'static>(
   });
   threads.insert(0, node_thread);
 
-  // Joins all threads
+  // Joins all threads and checks for panics.
   for thread in threads {
-    thread.join().unwrap();
+    match thread.join() {
+      Ok(_) => {
+        // For now at least, child threads are expected to handle their
+        // own errors, eg by logging.  We do not propagate them up.
+        // For a way to do it, see:
+        //   https://stackoverflow.com/a/62823352/10087197
+      }
+      Err(e) => {
+        // When join() returns an error, that means a child thread panicked.
+        let msg = if let Some(s) = e.downcast_ref::<&str>() {
+          format!("Child thread panicked with message:\n{}", s)
+        } else if let Some(s) = e.downcast_ref::<String>() {
+          format!("Child thread panicked with message:\n{}", s)
+        } else {
+          format!("Child thread panicked!\n{:?}", e)
+        };
+
+        // We return the panic error to propagate upwards so that
+        // panic in any thread will result in process termination
+        // as per rust default behavior.
+        // Alternatively we could just log the error and keep going
+        // without the thread that panicked, or even restart it.
+        return Err(anyhow!(msg));
```

Comment:
I think rust's default behavior is to continue running the program with the other threads that didn't panic, or am I wrong?

Reply:
It depends on if join() is used or not, and how it is used.

If thread::join().unwrap() is called and the child thread has panicked then the unwrap() in *our* thread will panic.  In this case, our thread is the main thread, so the process will exit as the panic unwinds.   This is the previous behavior I am preserving.   (although now we exit by propagating error up, not a panic).

If we do not call unwrap() then we can handle (or ignore) the error, in which case we continue running.

So I probably could word it better by saying "as per previous behavior" or "join.unwrap() behavior".

Now, a separate question is:  what behavior do we actually want?

We can think about:   what should happen if:
* an event thread panics
* a mining thread panics
* the api thread panics
* the node thread panics

We might choose a different answer for each.  I don't know.  That's something that can be decided later.

For this PR, I chose to keep the basic behavior the same:  by raising the error, main() will eventually return it and program exits.  (although with an error, not a panic)

And of course with these PRs, we do our best to avoid any of these threads panicking in the first place.  But we can't guarantee std lib or 3rd party crates won't.

Reply:
> We might choose a different answer for each. I don't know. That's something that can be decided later.

I get your point, and this is the best thing to do. I think I'll create an issue for this.